### PR TITLE
Initial YAML-based scipack

### PR DESCRIPTION
**Pull request by jdtsmith**
This is an initial implementation of the YAML-based science pack.  At present it is not coupled in any way to the model fitter, so details can be changes, do have a look.  Also included is the PAHFIT _classic_ model, encoded as `packs/classic.yml`.  The yaml file is read by a astropy.table.Table inheriting class (`Features`), so is a table in memory, similar to the current setup.  Note that the PAHFIT classic model fixes dust features (wavelength or FWHM).  Also note that value + bounds are grouped under a single column name as an `(n, 3)` column-array.  Going forward astropy 5.1 will have better formatting of these.

To test/play:

```python
from features import Features
t = Features.read('classic.yml')
```

### CODE REVIEW

**Comment by drvdputt** [comment]
After playing around, here are some tings that I found to be working or not working.

## What works as expected
comment out sub-features (e.g. some H2 line), e.g.
```
H2_lines:
    kind: line
    wavelength:
        H2_S(2):   12.2785
        # H2_S(1):   17.0346
        # H2_S(0):   28.2207
```
comment whole feature block (e.g. all H2 lines, a PAH feature)
```
# H2_lines:
#    kind: line
#    wavelength:
#        H2_S(7):    5.5115
#        H2_S(6):    6.1088
#       H2_S(5):    6.909107
```
Comment at end of temperature list e.g.
```
# Modified Blackbodies
    kind: dust_continuum
    temperature: [300, 200, 135, 90] #65, 50, 40, 35] 
```

Errors out correctly:
- add non-existing type of feature
- add wrong parameter name

## Cryptic or no error
**Errors out cryptically**
Comment out all H2 lines, but not the block itself (error happens in astropy column, during call to setting the format at features.py:289)
```
H2_lines:
    kind: line
    wavelength:
        # H2_S(2):   12.2785
        # H2_S(1):   17.0346
        # H2_S(0):   28.2207
```
Removing floating point number for one of the H2 wavelengths results in (TableMergeError: The 'wavelength' columns have incompatible types: ['object', 'float64’]).
```
H2_lines:
    kind: line
    wavelength:
        H2_S(2):  
        H2_S(1):   17.0346
        H2_S(0):   28.2207
```


**Silent failure**
Commenting out starlight / temperature results in table with blank temperature field
```
starlight:
    kind: starlight_continuum
    # temperature: 5000
```

**Comment by jdtsmith** [comment]
One other thing you may notice among the recent pushes is rolling back an attempt to use a structured array for bounded parameter columns.  I was hopeful this would work out after some discussions with astropy devs, but a few blocking bugs —astropy/astropy#13267, astropy/astropy#13271 — means this won't work for at least a couple point versions.  This would mostly be a formatting convenience so you could see `12.3 (12.28, 12.33)` for a bounded param when printing the table.  If you have any other clever ideas for this let me know.

**Comment by Dhruvil2911** [comment]
Just a minor error - The Fe[II] lines under the ionic lines in the classic.yml file (line number 37) need to have different keys (names), just as done for S[III] since YAML doesn't allow same names.

**Comment by jdtsmith** [comment]
> Just a minor error - The Fe[II] lines under the ionic lines in the classic.yml file (line number 37) need to have different keys (names), just as done for S[III] since YAML doesn't allow same names.

Thanks Dhruvil, fixed.  I also took the opportunity to patch the YAML loader so it throws an error if duplicate keys are used in one dict.

**Comment by jdtsmith** [comment]
For people trying to follow along or construct their own YAML-based scipacks, the draft docs are updated now: https://github.com/PAHFIT/pahfit/wiki/PAHFIT-2022#science-packs.

**Review by Ameek-Sidhu — COMMENTED** [review]
Other than the small error I found, everything works well!

File: pahfit/features/features.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+    parameter information.  Note that each parameter has an associated
+    `kind', and that each kind has an associated set of allowable
+    parameters (see _kind_params, below).
+    """
+
+    TableFormatter = BoundedParTableFormatter
+    MaskedColumn = BoundedMaskedColumn
+    
+    param_covar = TableAttribute(default=[])
+    _kind_params = {'starlight':           {'temperature',
```
[the remark this answers is not in the document]
Comment (Ameek-Sidhu) [reply pahfit/features/features.py:113]: There is one more issue I encountered. 
When I gave the `value` and `bounds` attributes to the `fwhm` parameter in `PAH_7.7_cmp`, it worked. 
However, when I do the same thing to the `fwhm` parameter in `PAH_10.7`, I get the following error:

> All non-group parameters in PAH_10.7 must be lists or dicts:

[the remark this answers is not in the document]
Comment (jdtsmith) [reply pahfit/features/features.py:113]: Can you show an example stanza that is failing?

File: pahfit/features/features.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+                    for (n,v) in elem['features'].items():
+                        if bounds and not 'bounds' in v: # inherit bounds
+                            v['bounds'] = bounds
+                        cls._add_feature(kind, feat_tables, n, group=name, **v)
+                elif hasLists:
+                    llen = []
+                    for k,v in elem.items():
+                        if k in cls._group_attrs: continue
+                        if not isinstance(v, (tuple,list,dict)):
+                            raise PAHFITFeatureError(f"All non-group parameters in {name} "
```
Comment (Ameek-Sidhu) [inline pahfit/features/features.py:229]: This part of the code raised the error.

Comment (jdtsmith) [reply pahfit/features/features.py:231]: Thanks @Ameek-Sidhu, I just uploaded a fix for this; can you try it out?

**Comment by jdtsmith** [comment]
> I agree to remove the FWHM for the unresolved lines and proceed as you propose. It doesn't make sense to keep it as it is set by the instrument pack.

Good; done.  Note that later we may add a "velocity_sigma" parameter or similar to lines, and dust_features, so as to allow them to be a "bit wider" than the instrument pack specifies. I guess this could be set on any individual or group of lines/dust_features.

**Comment by jdtsmith** [comment]
Fixes @Ameek-Sidhu's issue.  Tests for pack ingestion TBD with #218.

**Pull request merged**